### ISSUE TITLE
Saveable bots

### ISSUE THREAD

**Issue by musicman609** [MegaMek/megamek#2956]
### Environment
Ver 49, windows

### Description
Can there be a way to save a specific AI bot so that it can be loaded without having to enter in all the details manually? (Name, characteristics,  team, paint job,  etc)

This would be useful for long campaigns with multiple bots 

### Files
N/A

**Comment by SimonLandmine** [comment]
It would also be useful if bot configuration data was saved with a gamestate, so that if you have to stop a multiple-bot game partway through, and then reload it, you don't need to go through reconfiguring each bot.
Then, when you reload the game, you would only have to confirm the bot, rather than reconfigure values, retreat edges, etc.
(I usually use one bot per 'role', to ensure 'reasonable' Princess behaviour.)
That might be a separate RFE requirement, though.

**Comment by SJuliez** [comment]
The first issue is purely client-side and does not need to go to savegames. 
The second issue is adding knowledge about bot players to savegames which (I think) requires letting the server know that a connected player is a bot and what its settings are. As far as I know, this is currently not the case, not even the server knows if a connected player is a bot. (There is some wisdom in that as it avoids programming anything in the server in a way that assumes such knowledge about players which should be avoided). So these issues are indeed rather separate. Also I think that the savegame idea would be very helpful for games with an elaborate setup and 14 players when they have to restart for some reason.

**Comment by SimonLandmine** [comment]
Good points. I'll create a new RFE for that idea. I should have thought about the client/server separation.

**Comment by musicman609** [comment]
This way, if i wanted to set up a recurring bot for the Gray Death Legion, I could simply set it up how I want it, and then load it at need any time i use them as AIs in a campaign

**Comment by SimonLandmine** [comment]
Combining the bot presets (which are currently easy enough to load from the drop-down (and you can add new ones by editing the .xml)) with your preferred camo and name. What other variables would you want it to save?

**Comment by musicman609** [comment]
It might be reaching, but maybe attaching certain units (continuing with my prior example,  attaching the mechs I assign to 1st company of the Gray Death Legion) with that drop down would be nice... so that way, if I took the time to set it up once, I could load the gray death companions, with named/skilled pilots, their appropriate mechs, appropriate camo, and associated AI behavior all in one drop-down click would be the ultimate endgame of that I think.

**Comment by SimonLandmine** [comment]
So, basically embedding bot configurations in the saved unit file?

**Comment by musicman609** [comment]
Thats it!

I did see that I can alter the XML file, but I still have to name the unit, populate it, paint it, etc.

**Comment by SimonLandmine** [comment]
Yes, but once you've set up and customised the unit in the Lobby once (including setting 'individual camo' for all members, so that it's saved as part of the unit, you can save the whole lot from the lobby as an .mul file.
Then, when you need the unit again, you need only three simple steps ...
1. Load units from .mul file
2. Create bot player with deisred name, selecting your preferred (or custom) behaviour from the XML list.
3. Transfer units to bot.
And then you're ready to fight (once you've loaded your own units).
Your suggestion would combine steps 1 and 2.

**Comment by musicman609** [comment]
Oh, I save plenty of MULs, I just meant by the time I make sure everyone enters the map on the appropriate turn, the behaviors are right,  everyone's on the proper team,  proper camo, etc its a lot of boxes to check 

It isn't a complaint, I *love* the scope of all the details we can change and modify... I just thought it could be a good way to streamline things!

**Comment by SimonLandmine** [comment]
That's cool - just making sure that you were aware of MULs! The request still makes sense - "OpFor-in-a-can"!

**Comment by SJuliez** [comment]
Actually you can create the bot, select it and load any MULs directly to it (by having selected it before, no need to transfer units after loading).

**Comment by SimonLandmine** [comment]
Ah, I didn't know you could do that. [makes note]
So you'd just need to ...
1. Create bot with presaved set of settings. (Which you can do by editing the .xml)
2. With it selected, load an MUL (which would have the units and their individual camouflage)
Nice!

**Comment by SJuliez** [comment]
Unless I missed something, editing the xml filte isn't necessary. Enter the behavior and give it a name in the presets dropdown (it is an editable text field as well). When you add the bot, you will be asked if you want to save that preset.

**Comment by SimonLandmine** [comment]
I don't think that's commonly known! I didn't know it either. :-)

**Comment by musicman609** [comment]
You can name the behavior, but ive yet to see how i can automatically name the bot. I can select a named behavior from the drop down list, but it will still default to the name of "Princess" at the bottom

**Comment by SimonLandmine** [comment]
You can edit that field when you're configuring a newly created bot instance, rather than setting it up in the behaviour.

**Comment by musicman609** [comment]
Can you set up camo too?

**Comment by SimonLandmine** [comment]
You can't save a camo as part of a Princess configuration, but you can save a camo with the units in a MUL, if you set it to 'individual camo' before saving the MUL. So you can load a MUL with saved camo, and then load a saved Princess to control it.

**Comment by rjhancock** [comment]
Adding this to my larger project as an additional file format to be saved.

**Comment by HammerGS** [comment]
Some of this now works, so it is worth recording where the line falls.

What already persists: a bot's behaviour settings survive a save and reload. The game keeps each bot's settings by player name and, alongside them, which kind of bot last held that seat, and both go into the savegame. Reloading restores every bot to the settings and the AI type it had rather than guessing. The Edit Bots dialog shows a Restore button on any seat that has a saved configuration waiting. Separately, Princess behaviour profiles can be saved under a name and reused between games from the bot configuration dialog.

What is still missing is the rest of what this request asks for: the bot's name, its team and its camouflage are not part of any of that. Setting up the same opposing force across a long campaign still means re-entering those by hand each time.

Labelled GM Tools and left open for the remaining part.

**Comment by HammerGS** [comment]
Status note, so anyone picking this up knows where the Game Master work stands.

A run of Game Master tooling has just gone up for review, covering three of the requests in this area:

- **#8785** - editing terrain and buildings during a game (PR #8803)
- **#8786** - adding units to another player from the lobby (PR #8808)
- **#8807** - bringing a player into a game already in progress: team, deployment edge and a force (PR #8813, stacked on #8808)

**This issue is not part of that work and remains open.** It is being left for a later pass rather than dropped.

Two findings from building the above are worth recording here, because both affect this area and neither is fixed by those pull requests:

- Whether a deployment phase happens at all is decided by an exact match on the arrival round, while a unit reports itself ready on or before its round. A unit whose round passes while its owner still cannot deploy is therefore never called for again. The new tooling makes that unreachable through the interface and logs it when it happens, but the mismatch itself is untouched and belongs in its own change.
- `receiveEntityAdd` does not check that the sender owns the units being added; it trusts the owner id in the packet. That is long-standing behaviour and is what makes reinforcing another player work at all, but it also means any client can add units to any player.
